### PR TITLE
Added the ability to evaluate linear recurrences without obtaining closed form expressions

### PR DESCRIPTION
**Pull request by sidhantnagpal**
#### Brief description of what is fixed or changed
`linrec(coeffs, init, n)` takes the coefficients, initial values, and point of evaluation of the recurrence as input and produces the corresponding result.
Usage for a recurrence like `f(n) = f(n - 7) + f(n - 13) + f(n - 17)` (having order `17`) would be:
```
In []: coeffs, init = [0]*17, [1]*17
In []: coeffs[7 - 1] = coeffs[13 - 1] = coeffs[17 - 1] = 1
In []: [linrec(coeffs, init, n) for n in range(40, 50)]
Out[]: [17, 21, 21, 23, 29, 31, 31, 35, 41, 41]

In []: def f(n):
  ...:     if n < 17:
  ...:         return 1
  ...:     return f(n - 7) + f(n - 13) + f(n - 17)
  ...: 

In []: [f(n) for n in range(40, 50)]
Out[]: [17, 21, 21, 23, 29, 31, 31, 35, 41, 41]
```

#### Other comments

### CODE REVIEW

**Comment by asmeurer** [comment]
This seems like it should be an internal function (not in `__init__.py`) and the user-facing function should accept a symbolic input (like the same format as `rsolve`), and work even for nonlinear recurrences.

**Comment by sidhantnagpal** [comment]
Yes, that would be better. `NotImplementedError` could be raised for the cases not covered by this approach (I am not sure if it's possible to do that for nonlinear case).

File: sympy/discrete/recurrence.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+    def _final_coeffs(n):
+        # computes the final coefficient list - `cf` corresponding to the
+        # point at which recurrence is to be evalauted - `n`, such that,
+        # y(n) = cf[0]*y(k - 1) + cf[1]*y(k - 2) + ... + cf[k - 1]*y(0)
+
+        if n < k:
+            return [S.Zero]*n + [S.One] + [S.Zero]*(k - n - 1)
+        else:
+            return _square_and_reduce(_final_coeffs(n // 2), offset=n%2)
```
Comment (jksuom) [inline sympy/discrete/recurrence.py]: Is the parameter `offset` explicitly written for a particular reason?

**Comment by asmeurer** [comment]
Or it could just evaluate it directly (but NotImplementedError is also fine for now).

File: sympy/discrete/recurrence.py (hunk 1 above)
Comment (sidhantnagpal) [reply sympy/discrete/recurrence.py]: It can be dropped, there's no specific reason for writing it.

File: sympy/discrete/recurrence.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    ==========
+
+    coeffs : iterable
+        coefficients of recurrence
+    init : iterable
+        initial values of recurrence
+    n : Integer
+        point of evaluation of recurrence
+
+    Explanation
```
Comment (jksuom) [inline sympy/discrete/recurrence.py]: The name of this section should probably be `Notes`. See https://numpydoc.readthedocs.io/en/latest/format.html#sections.

File: sympy/discrete/recurrence.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    of the recurrence if and only if `T(x^i p(x)) = 0` for each `i ge 0` where
+    `p(x) = x^k - c_0 x^(k-1) - \cdots - c_{k-1}` is the characteristic
+    polynomial.
+
+    Then `T(f(x)p(x)) = 0` for each polynomial `f(x)` (as it is a linear
+    combination of powers `x^i`). Now, if `x^n` is congruent to
+    `g(x) = a_0 x^0 + a_1 x^1 + \cdots + a_{k-1} x^{k-1}` modulo `p(x)`, then
+    `T(x^n) = x_n` is equal to
+    `T(g(x)) = a_0 x_0 + a_1 x_1 + \cdots + a_{k-1} x_{k-1}`.
+
```
Comment (jksuom) [inline sympy/discrete/recurrence.py]: A note on the computation of `x^n` by squaring could go here.

**Comment by jksuom** [comment]
Thanks. We can return to the public recursion API later.

**Pull request merged**